### PR TITLE
Added add and remove root keys to cmd

### PR DESCRIPTION
**Pull request by neilnaveen**
- Added functionality to cmd to add and remove root keys from, which needs a key already in the root to allow adding or removing keys from the root.
- Partially fixes issue https://github.com/gittuf/gittuf/issues/117

### CODE REVIEW

File: internal/policy/root.go [hunk 1]
```diff
@@ -27,6 +27,49 @@ func InitializeRootMetadata(key *tuf.Key) *tuf.RootMetadata {
 	return rootMetadata
 }
 
+func AddRootKey(rootMetadata *tuf.RootMetadata, rootKey *tuf.Key) *tuf.RootMetadata {
+	rootMetadata.Keys[rootKey.KeyID] = rootKey
+	if _, ok := rootMetadata.Roles[RootRoleName]; !ok {
+		return InitializeRootMetadata(rootKey)
```
Comment (adityasaky) [inline internal/policy/root.go:33]: hm, I don't know that we want to initialize the root role if it doesn't exist. Let's keep this separate for now, I think it's worth exploring this in the context of #45.

File: internal/policy/root.go [hunk 2]
```diff
@@ -27,6 +27,49 @@ func InitializeRootMetadata(key *tuf.Key) *tuf.RootMetadata {
...
+	}
+
+	rootMetadata.AddKey(rootKey)
+
+	rootRoles := rootMetadata.Roles[RootRoleName]
```
Comment (adityasaky) [inline internal/policy/root.go:38]:
```suggestion
	rootRole := rootMetadata.Roles[RootRoleName]
```

File: internal/policy/root.go [hunk 3]
```diff
@@ -27,6 +27,49 @@ func InitializeRootMetadata(key *tuf.Key) *tuf.RootMetadata {
...
+
+	for _, keyID := range rootRoles.KeyIDs {
+		if keyID == rootKey.KeyID {
+			return rootMetadata
+		}
+	}
+
+	rootRoles.KeyIDs = append(rootRoles.KeyIDs, rootKey.KeyID)
+
```
Comment (adityasaky) [inline internal/policy/root.go:47]:
```suggestion
```

File: internal/policy/root.go [hunk 4]
```diff
@@ -27,6 +27,49 @@ func InitializeRootMetadata(key *tuf.Key) *tuf.RootMetadata {
 	return rootMetadata
 }
 
+func AddRootKey(rootMetadata *tuf.RootMetadata, rootKey *tuf.Key) *tuf.RootMetadata {
+	rootMetadata.Keys[rootKey.KeyID] = rootKey
+	if _, ok := rootMetadata.Roles[RootRoleName]; !ok {
+		return InitializeRootMetadata(rootKey)
+	}
+
+	rootMetadata.AddKey(rootKey)
```
Comment (adityasaky) [inline internal/policy/root.go:36]: isn't this fundamentally the same as line 31? I suggest removing line 31 in that case.

File: internal/policy/root.go [hunk 5]
```diff
@@ -27,6 +27,49 @@ func InitializeRootMetadata(key *tuf.Key) *tuf.RootMetadata {
 	return rootMetadata
 }
 
+func AddRootKey(rootMetadata *tuf.RootMetadata, rootKey *tuf.Key) *tuf.RootMetadata {
```
Comment (adityasaky) [inline internal/policy/root.go:30]: missing docstring

File: internal/policy/root.go [hunk 6]
```diff
@@ -27,6 +27,49 @@ func InitializeRootMetadata(key *tuf.Key) *tuf.RootMetadata {
...
+	}
+
+	rootRoles.KeyIDs = append(rootRoles.KeyIDs, rootKey.KeyID)
+
+	rootMetadata.Roles[RootRoleName] = rootRoles
+
+	return rootMetadata
+}
+
+func RemoveRootKey(rootMetadata *tuf.RootMetadata, keyID string) (*tuf.RootMetadata, error) {
```
Comment (adityasaky) [inline internal/policy/root.go:53]: missing docstring

Comment (adityasaky) [inline internal/policy/root.go:53]: For consistency with targets key, I wonder if this should be `Delete`.

File: internal/policy/root_test.go [hunk 7]
```diff
@@ -29,6 +29,66 @@ func TestInitializeRootMetadata(t *testing.T) {
 	assert.Equal(t, []string{key.KeyID}, rootMetadata.Roles[RootRoleName].KeyIDs)
 }
 
+func TestAddRootKey(t *testing.T) {
+	keyBytes, err := os.ReadFile(filepath.Join("test-data", "root.pub"))
```
Comment (adityasaky) [inline internal/policy/root_test.go:33]: https://github.com/gittuf/gittuf/blob/main/internal/policy/helpers_test.go#L23-L36

feel free to use these with tuf.LoadKey... directly.

File: internal/policy/root_test.go [hunk 8]
```diff
@@ -29,6 +29,66 @@ func TestInitializeRootMetadata(t *testing.T) {
...
+	}
+
+	rootMetadata = AddRootKey(rootMetadata, newRootKey)
+
+	rootMetadata, err = RemoveRootKey(rootMetadata, newRootKey.KeyID)
+
+	assert.Nil(t, err)
+	assert.Equal(t, key, rootMetadata.Keys[key.KeyID])
+	assert.Equal(t, newRootKey, rootMetadata.Keys[newRootKey.KeyID])
+	assert.Equal(t, []string{key.KeyID}, rootMetadata.Roles[RootRoleName].KeyIDs)
```
Comment (adityasaky) [inline internal/policy/root_test.go:89]: Can you try removing the original key too and trigger ErrCannotMeetThreshold?

File: internal/repository/root_test.go [hunk 9]
```diff
@@ -43,6 +44,116 @@ func TestInitializeRoot(t *testing.T) {
 	assert.Nil(t, err)
 }
 
+func TestAddRootKey(t *testing.T) {
+	r, keyBytes := createTestRepositoryWithRoot(t, "")
+
+	key, err := tuf.LoadKeyFromBytes(keyBytes)
```
Comment (adityasaky) [inline internal/repository/root_test.go:50]: feel free to set up a helpers_test.go file in this package for the `// embed` directives, similar to the policy package. https://github.com/gittuf/gittuf/blob/main/internal/repository/targets_test.go#L20 has a single directive right now.

File: internal/cmd/trust/addrootkey/addrootkey.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+	return repo.AddRootKey(cmd.Context(), rootKeyBytes, newRootKeyBytes, true)
+}
+
+func New(persistent *persistent.Options) *cobra.Command {
+	o := &options{p: persistent}
+	cmd := &cobra.Command{
+		Use:     "add-root-key",
+		Short:   "Add Root key to gittuf root of trust",
+		Long:    `This command allows users to add a new trusted key to the root. Note that authorized keys can be specified from disk using the custom securesystemslib format, from the GPG keyring using the "gpg:<fingerprint>" format, or as a Sigstore identity as "fulcio:<identity>::<issuer>".`,
```
Comment (adityasaky) [inline internal/cmd/trust/addrootkey/addrootkey.go:53]: This is currently incorrect, something we ought to fix in addpolicykey too until #229 is fixed. Or at the very least, these can be specified and it'll work but signing the metadata will fail.

maybe we can leave it in and focus on that through #229 instead.

File: internal/cmd/trust/removerootkey/removerootkey.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+	"github.com/gittuf/gittuf/internal/cmd/common"
+	"github.com/gittuf/gittuf/internal/cmd/trust/persistent"
+	"github.com/gittuf/gittuf/internal/repository"
+	"github.com/spf13/cobra"
+)
+
+type options struct {
+	p            *persistent.Options
+	newRootKeyID string
```
Comment (adityasaky) [inline internal/cmd/trust/removerootkey/removerootkey.go:17]:
```suggestion
	rootKeyID string
```

this isn't a new key anymore.

File: internal/cmd/trust/trust.go [hunk 12]
```diff
@@ -22,6 +24,8 @@ func New() *cobra.Command {
 	cmd.AddCommand(i.New(o))
 	cmd.AddCommand(addpolicykey.New(o))
 	cmd.AddCommand(removepolicykey.New(o))
+	cmd.AddCommand(addrootkey.New(o))
```
Comment (adityasaky) [inline internal/cmd/trust/trust.go:27]: nit: can you sort these alphabetically by the subcommand's package name? It wasn't originally the case but as this grows, let's keep them organized.

File: internal/repository/root.go [hunk 13]
```diff
@@ -53,6 +53,109 @@ func (r *Repository) InitializeRoot(ctx context.Context, rootKeyBytes []byte, si
 	return state.Commit(ctx, r.r, commitMessage, signCommit)
 }
 
+func (r *Repository) AddRootKey(ctx context.Context, rootKeyBytes, newrootKeyBytes []byte, signCommit bool) error {
```
Comment (adityasaky) [inline internal/repository/root.go:56]: missing docstring

File: internal/repository/root.go [hunk 14]
```diff
@@ -53,6 +53,109 @@ func (r *Repository) InitializeRoot(ctx context.Context, rootKeyBytes []byte, si
...
+	}
+
+	state.RootEnvelope = env
+
+	commitMessage := fmt.Sprintf("Add root key '%s' to root", newRootKey.KeyID)
+
+	return state.Commit(ctx, r.r, commitMessage, signCommit)
+}
+
+func (r *Repository) RemoveRootKey(ctx context.Context, rootKeyBytes []byte, keyID string, signCommit bool) error {
```
Comment (adityasaky) [inline internal/repository/root.go:108]: missing docstring

**Review by adityasaky — CHANGES_REQUESTED** [review]
This is largely in good shape, I think it'll be good to go once these comments are addressed. Thanks for working on this!

File: internal/policy/root.go (hunk 1 above)
Comment (neilnaveen) [reply internal/policy/root.go]: Now AddRootKey just returns the given rootMetadata if the root role does not exist

File: internal/policy/root.go (hunk 2 above)
Comment (neilnaveen) [reply internal/policy/root.go]: Fixed

File: internal/policy/root.go (hunk 3 above)
Comment (neilnaveen) [reply internal/policy/root.go]: Fixed

File: internal/policy/root.go (hunk 5 above)
Comment (neilnaveen) [reply internal/policy/root.go:32]: Added

File: internal/policy/root.go (hunk 6 above)
Comment (neilnaveen) [reply internal/policy/root.go]: Added

Comment (neilnaveen) [reply internal/policy/root.go]: Changed

File: internal/policy/root_test.go (hunk 7 above)
Comment (neilnaveen) [reply internal/policy/root_test.go]: I have switched it to use the embedded variables

File: internal/policy/root_test.go (hunk 8 above)
Comment (neilnaveen) [reply internal/policy/root_test.go:71]: Added

File: internal/repository/root_test.go (hunk 9 above)
Comment (neilnaveen) [reply internal/repository/root_test.go:48]: I have added a helper_test.go file, with all test repo creation functions, and all embedded variables in it

File: internal/cmd/trust/addrootkey/addrootkey.go (hunk 10 above)
Comment (neilnaveen) [reply internal/cmd/trust/addrootkey/addrootkey.go:53]: Should I just take out the last sentence of the long --help for add root key?

File: internal/cmd/trust/removerootkey/removerootkey.go (hunk 11 above)
Comment (neilnaveen) [reply internal/cmd/trust/removerootkey/removerootkey.go]: Fixed

File: internal/cmd/trust/trust.go (hunk 12 above)
Comment (neilnaveen) [reply internal/cmd/trust/trust.go:26]: Fixed

File: internal/repository/root.go (hunk 13 above)
Comment (neilnaveen) [reply internal/repository/root.go:58]: Added

File: internal/repository/root.go (hunk 14 above)
Comment (neilnaveen) [reply internal/repository/root.go:112]: Added

File: internal/cmd/trust/addrootkey/addrootkey.go (hunk 10 above)
Comment (adityasaky) [reply internal/cmd/trust/addrootkey/addrootkey.go:53]: for now, yes.

File: internal/policy/root_test.go [hunk 15]
```diff
@@ -29,56 +29,84 @@ func TestInitializeRootMetadata(t *testing.T) {
...
 	}
 
-	targetsKey1, err := tuf.LoadKeyFromBytes(targetsKeyBytes)
+	rootMetadata := InitializeRootMetadata(key)
+
+	targetsKey1, err := tuf.LoadKeyFromBytes(targets1KeyBytes)
 	if err != nil {
 		t.Fatal(err)
 	}
+	var targetsKeyBytes []byte
```
Comment (adityasaky) [inline internal/policy/root_test.go:109]:
```suggestion
```

I'd just add make it `:=` below instead.

Comment (adityasaky) [reply internal/policy/root_test.go:109]: Alternatively, that can be added to helpers_test via go:embed too.

File: internal/repository/helpers_test.go [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+	authorizedKeys := [][]byte{kb}
+
+	if err := r.AddDelegation(testCtx, targetsKeyBytes, policy.TargetsRoleName, "protect-main", authorizedKeys, []string{"git:refs/heads/main"}, false); err != nil {
+		t.Fatal(err)
+	}
+
+	return r
+}
+
+func createTestRepositoryWithTargets(t *testing.T) (*Repository, []byte) {
```
Comment (adityasaky) [inline internal/repository/helpers_test.go:88]: Slightly unrelated, but can you check if this can be removed altogether in favor of createTestRepositoryWithPolicy? I see this returns pub key bytes but we should have that from the embed anyway...

File: internal/repository/root.go [hunk 17]
```diff
@@ -53,6 +53,113 @@ func (r *Repository) InitializeRoot(ctx context.Context, rootKeyBytes []byte, si
...
+
+	rootMetadata, err := state.GetRootMetadata()
+	if err != nil {
+		return err
+	}
+
+	if !isKeyAuthorized(rootMetadata.Roles[policy.RootRoleName].KeyIDs, rootKeyID) {
+		return ErrUnauthorizedKey
+	}
+	newRootKey, err := tuf.LoadKeyFromBytes(newrootKeyBytes)
```
Comment (adityasaky) [inline internal/repository/root.go:81]:
```suggestion

	newRootKey, err := tuf.LoadKeyFromBytes(newrootKeyBytes)
```

File: internal/repository/root.go [hunk 18]
```diff
@@ -53,6 +53,113 @@ func (r *Repository) InitializeRoot(ctx context.Context, rootKeyBytes []byte, si
...
+
+	state.RootEnvelope = env
+
+	commitMessage := fmt.Sprintf("Add root key '%s' to root", newRootKey.KeyID)
+
+	return state.Commit(ctx, r.r, commitMessage, signCommit)
+}
+
+// RemoveRootKey is the interface for the user to de-authorize a key
+// trusted to sign the Root role / policy file.
```
Comment (adityasaky) [inline internal/repository/root.go:111]:
```suggestion
// trusted to sign the Root role.
```

File: internal/repository/root_test.go [hunk 19]
```diff
@@ -43,6 +42,106 @@ func TestInitializeRoot(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	sv, err := signerverifier.NewSignerVerifierFromSecureSystemsLibFormat(keyBytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var newRootKey *sslibsignerverifier.SSLibKey
+
+	newRootKey, err = tuf.LoadKeyFromBytes(targetsKeyBytes)
```
Comment (adityasaky) [inline internal/repository/root_test.go:59]:
```suggestion
	newRootKey, err := tuf.LoadKeyFromBytes(targetsKeyBytes)
```

This can now remove the `newRootKey` declaration right?

File: internal/repository/root_test.go [hunk 20]
```diff
@@ -43,6 +42,106 @@ func TestInitializeRoot(t *testing.T) {
...
+	rootKey, err := tuf.LoadKeyFromBytes(keyBytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+	sv, err := signerverifier.NewSignerVerifierFromSecureSystemsLibFormat(keyBytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = r.AddRootKey(context.Background(), keyBytes, keyBytes, false)
```
Comment (adityasaky) [inline internal/repository/root_test.go:94]: Interesting that this is actually a noop in terms of the modification to the root role's contents, but the version is still bumped and a new state is created... not a problem for now, and perhaps not a real problem at all.

File: internal/repository/root_test.go [hunk 21]
```diff
@@ -43,6 +42,106 @@ func TestInitializeRoot(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	rootMetadata, err := state.GetRootMetadata()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, 3, rootMetadata.Version)
+	assert.Equal(t, rootKey.KeyID, rootMetadata.Roles[policy.RootRoleName].KeyIDs[0])
```
Comment (adityasaky) [inline internal/repository/root_test.go:120]: This check can be removed given the `Contains` checks below.

**Review by adityasaky — CHANGES_REQUESTED** [review]
This one's almost there!

File: internal/repository/root_test.go (hunk 21 above)
Comment (neilnaveen) [reply internal/repository/root_test.go]: Removed

File: internal/repository/root_test.go (hunk 19 above)
Comment (neilnaveen) [reply internal/repository/root_test.go:59]: Removed

File: internal/repository/root.go (hunk 18 above)
Comment (neilnaveen) [reply internal/repository/root.go]: Removed

File: internal/repository/root.go (hunk 17 above)
Comment (neilnaveen) [reply internal/repository/root.go:82]: Removed

File: internal/policy/root_test.go (hunk 15 above)
Comment (neilnaveen) [reply internal/policy/root_test.go]: removed

File: internal/repository/sync_test.go [hunk 22]
```diff
@@ -196,7 +183,7 @@ func TestClone(t *testing.T) {
 		defer os.Chdir(currentDir) //nolint:errcheck
 
 		dirName := "myRepo"
-		if err := os.Mkdir(dirName, 0755); err != nil {
+		if err := os.Mkdir(dirName, 0o755); err != nil {
```
Comment (adityasaky) [inline internal/repository/sync_test.go:186]: unrelated change? `0755` is already octal no? Looks like go supports `0`, `0o`, and `0O`: https://go.dev/ref/spec#Integer_literals

File: internal/repository/helpers_test.go (hunk 16 above)
Comment (adityasaky) [reply internal/repository/helpers_test.go:88]: Let's handle this in a separate PR.

**Review by adityasaky — APPROVED** [review]
LGTM, bt can you:
a) reverse the octal literal in `os.Mkdir`; and
b) clean up commit history?

Thanks!

File: internal/cmd/trust/addrootkey/addrootkey.go (hunk 10 above)
Comment (adityasaky) [reply internal/cmd/trust/addrootkey/addrootkey.go:53]: Also bump on this one.

**Comment by neilnaveen** [comment]
> fix CLI doc until we unify signing methods (https://github.com/gittuf/gittuf/issues/229)

What do you mean? How should I change the docs, should it take in account unifying signing methods, so when they come, the doc is already right but wrong for now, or something else?

File: internal/repository/sync_test.go (hunk 22 above)
Comment (neilnaveen) [reply internal/repository/sync_test.go:186]: This was a typo, did not mean to change.

**Comment by adityasaky** [comment]
> What do you mean? How should I change the docs, should it take in account unifying signing methods, so when they come, the doc is already right but wrong for now, or something else?

Just this: https://github.com/gittuf/gittuf/pull/225#discussion_r1437144988

Let's get that line out for now.

**Comment by neilnaveen** [comment]
> > What do you mean? How should I change the docs, should it take in account unifying signing methods, so when they come, the doc is already right but wrong for now, or something else?
> 
> Just this: [#225 (comment)](https://github.com/gittuf/gittuf/pull/225#discussion_r1437144988)
> 
> Let's get that line out for now.

Done

File: internal/repository/helpers_test.go [hunk 23]
```diff
@@ -0,0 +1,103 @@
+package repository
```
Comment (adityasaky) [inline internal/repository/helpers_test.go:1]: nit: SPDX identifier

**Review by adityasaky — APPROVED** [review]
LGTM! Can you address the nit comment and squash your commits? Thanks!

**Review by adityasaky — APPROVED** [review]
